yew-functional/src/lib.rs

```diff
@@ -173,204 +169,76 @@ where
-    );
 }
 
-pub fn use_effect_with_deps<F, Destructor, Dependents>(callback: F, deps: Dependents)
-where
-    F: FnOnce(&Dependents) -> Destructor + 'static,
-    Destructor: FnOnce() + 'static,
-    Dependents: PartialEq + 'static,
-{
-    struct UseEffectState<Dependents, Destructor> {
-        deps: Rc<Dependents>,
-        destructor: Option<Box<Destructor>>,
-    }
-    impl<Dependents, Destructor: FnOnce() + 'static> Hook for UseEffectState<Dependents, Destructor> {
-        fn tear_down(&mut self) {
-            if let Some(destructor) = self.destructor.take() {
-                destructor()
-            }
-        }
-    }
-
-    let deps = Rc::new(deps);
-    let deps_c = deps.clone();
-
-    use_hook(
-        move |_state: &mut UseEffectState<Dependents, Destructor>, hook_callback| {
-            hook_callback(
-                move |state: &mut UseEffectState<Dependents, Destructor>| {
-                    if state.deps != deps {
-                        if let Some(de) = state.destructor.take() {
-                            de();
-                        }
-                        let new_destructor = callback(deps.borrow());
-                        state.deps = deps;
-                        state.destructor.replace(Box::new(new_destructor));
-                    } else if state.destructor.is_none() {
-                        state
-                            .destructor
-                            .replace(Box::new(callback(state.deps.borrow())));
-                    }
-                    false
-                },
-                true, // run post render
-            );
-        },
-        || UseEffectState {
-            deps: deps_c,
-            destructor: None,
-        },
-    );
+pub trait Hook {
+    type Output;
+    type Args;
+    fn tear_down(&mut self) {}
+    fn runner(&mut self, args: Self::Args, updater: HookUpdater) -> Self::Output;
 }
 
-pub fn use_hook<InternalHookState, HookRunner, R, InitialStateProvider, HookUpdate: 'static>(
-    hook_runner: HookRunner,
-    initial_state_producer: InitialStateProvider,
-) -> R
-where
-    HookRunner: FnOnce(&mut InternalHookState, Box<dyn Fn(HookUpdate, bool)>) -> R,
-    InternalHookState: Hook + 'static,
-    InitialStateProvider: FnOnce() -> InternalHookState,
-    HookUpdate: FnOnce(&mut InternalHookState) -> bool,
-{
+pub fn use_hook<InternalHook: Hook + 'static, I: FnOnce() -> InternalHook>(
+    args: InternalHook::Args,
+    initializer: I,
+) -> InternalHook::Output {
     // Extract current hook
-    let (hook, process_message) = CURRENT_HOOK.with(|hook_state_holder| {
-        let hook_state_holder = hook_state_holder.try_borrow_mut();
-        let mut hook_state_holder = hook_state_holder.expect("Nested hooks not supported");
+    let updater = CURRENT_HOOK.with(|hook_state_holder| {
+        let mut hook_state_holder = hook_state_holder
+            .try_borrow_mut()
+            .expect("Nested hooks not supported");
```

Comment:
Did the previous version supported hooks? It's pretty important for custom hooks.

Reply:
It's possible to create custom functions that leverage built-in hooks, but it is not possible to use the primitive "use_hook" in a nested context. However, I don't see how the API would suggest that this is even possible.